### ISSUE TITLE
Undefined behavior after `recent_entries_map_test`

### ISSUE THREAD

**Issue by kbr-scylla** [scylladb/scylladb#13128]
Observed on `master` build job on aarch64 https://jenkins.scylladb.com/job/scylla-master/job/build/1389/artifact/testlog/aarch64/debug/boost.recent_entries_map_test.test_update_and_remove_least_recent_entries.1.log/*view*/

full log is
```
=== TEST.PY STARTING TEST boost.recent_entries_map_test.test_update_and_remove_least_recent_entries.1 ===
export UBSAN_OPTIONS='halt_on_error=1:abort_on_error=1:suppressions=/jenkins/workspace/scylla-master/build/scylla/ubsan-suppressions.supp'
export ASAN_OPTIONS='disable_coredump=0:abort_on_error=1:detect_stack_use_after_return=1'
build/debug/test/boost/recent_entries_map_test --run_test=test_update_and_remove_least_recent_entries --report_level=no --logger=HRF,test_suite:XML,test_suite,/jenkins/workspace/scylla-master/build/scylla/testlog/aarch64/debug/xml/boost.recent_entries_map_test.test_update_and_remove_least_recent_entries.1.xunit.xml --catch_system_errors=no --color_output=false -- -c2 -m2G --overprovisioned --unsafe-bypass-fsync 1 --kernel-page-cache 1 --blocked-reactor-notify-ms 2000000 --collectd 0 --max-networking-io-control-blocks=100
=== TEST.PY TEST boost.recent_entries_map_test.test_update_and_remove_least_recent_entries.1 OUTPUT ===
Running 1 test case...
Entering test module "test/boost/recent_entries_map_test.cc"
test/boost/recent_entries_map_test.cc(0): Test case "test_entry_insertion" is skipped because disabled
test/boost/recent_entries_map_test.cc(0): Test case "test_entry_removal_by_key" is skipped because disabled
test/boost/recent_entries_map_test.cc(0): Test case "test_remove_least_recent_entries" is skipped because disabled
test/boost/recent_entries_map_test.cc(0): Entering test case "test_update_and_remove_least_recent_entries"
WARNING: debug mode. Not for benchmarking or production
INFO  2023-03-09 13:31:49,024 seastar - Reactor backend: linux-aio
WARN  2023-03-09 13:31:49,026 [shard 0] seastar - Creation of perf_event based stall detector failed, falling back to posix timer: std::system_error (error system:13, perf_event_open() failed: Permission denied)
INFO  2023-03-09 13:31:49,034 [shard 0] seastar - Created fair group io-queue-0, capacity rate 2147483:2147483, limit 12582912, rate 16777216 (factor 1), threshold 2000
INFO  2023-03-09 13:31:49,034 [shard 0] seastar - IO queue uses 0.75ms latency goal for device 0
INFO  2023-03-09 13:31:49,035 [shard 0] seastar - Created io group dev(0), length limit 4194304:4194304, rate 2147483647:2147483647
INFO  2023-03-09 13:31:49,035 [shard 0] seastar - Created io queue dev(0) capacities: 512:2000:2000 1024:3000:3000 2048:5000:5000 4096:9000:9000 8192:17000:17000 16384:33000:33000 32768:65000:65000 65536:129000:129000 131072:257000:257000
WARN  2023-03-09 13:31:49,036 [shard 1] seastar - Creation of perf_event based stall detector failed, falling back to posix timer: std::system_error (error system:13, perf_event_open() failed: Permission denied)
random-seed=3204665054
==12007==WARNING: ASan doesn't fully support makecontext/swapcontext functions and may produce false positives in some cases!
test/boost/recent_entries_map_test.cc(0): Leaving test case "test_update_and_remove_least_recent_entries"; testing time: 668018us
Leaving test module "test/boost/recent_entries_map_test.cc"; testing time: 668102us
/jenkins/workspace/scylla-master/build/scylla/seastar/src/core/reactor.cc:2927:25: runtime error: member access within null pointer of type 'seastar::reactor::task_queue'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /jenkins/workspace/scylla-master/build/scylla/seastar/src/core/reactor.cc:2927:25 in 
Aborting on shard 0.
Backtrace:
  0xa8f3c97
  0xffffb875e623
  0xffffb875e207
  0xffffb856d7eb
  0xffffb859bdb3
  0xffffb86b6ac7
  0xffffb86b6cc7
  0xffffb86b6b53
  linux-vdso.so.1+0x7db
  /lib64/libc.so.6+0x86c27
  /lib64/libc.so.6+0x3edff
  /lib64/libc.so.6+0x2b283
  0xa959417
  0xa95710b
  0xa96ee33
  0xa9717d7
  0xa972167
  0xffffb8594757
  0xffffb85aada7
  0xffffb819e7b7
  0xffffb7cf639f
  0xffffb7ddd38f
  0xffffb7db27d3
  0xffffb894c95b
  0xffffb894c9db
  0xffffb86b5c93
  0xffffb86b5b13
  0xffffb86b57b3
  0xffffb86b4d37
  0xffffb85a45f7
  0xffffb85a4493
  0xffffb85a2237
  0xffffb85b53ab
  0xffffb8070763
  0xffffb8070557
  0xffffb80704e7
  0xffffb807047b
  0xffffb807033f
  0xffffb806e87f
  0xb0b562f
  0xb0b555b
  0xb0b54bb
  0xb0b4eef
  0xffffb805f36b
  0xffffb801e61b
  0xffffb85641a3
  0xffffb89b1e47
  0xffffb872ce03
  /lib64/libc.so.6+0x412d7
  /lib64/libc.so.6+0x84ea3
  /lib64/libc.so.6+0xf07db
=== TEST.PY SUMMARY START ===
Test exited with code -6

=== TEST.PY SUMMARY END ===
```

Curious that UB was reported in... seastar reactor?
But I'm guessing that the recently added `recent_entries_map` data structure somehow caused it, since it happened in after leaving `recent_entries_map_test` (after `test_update_and_remove_least_recent_entries`).

**Comment by kbr-scylla** [comment]
I'm not sure if there's a way to render the stacktrace - IIUC we'd need access to the boost test executable... @bhalevy?

**Comment by andriipatsula** [comment]
Hello @kbr-scylla , I'll check tomorrow and/or during weekend. Do you have any suggestion how can I simulate (repeat locally) the problem? I may try to run the specific set of tests in a loop for N times in debug mode...

**Comment by kbr-scylla** [comment]
After compiling
```
./configure.py --mode debug
ninja -j<cores> build/debug/test/boost/recent_entries_map_test
```

you can do
```
./test.py --mode debug recent_entries_map_test --repeat <repeats>
```

Other than that, I don't know. It might be specific to ARM only (I'm not aware of this happening in other builds) if there's some implementation dependent code...

**Comment by bhalevy** [comment]
> I'm not sure if there's a way to render the stacktrace - IIUC we'd need access to the boost test executable... @bhalevy?

Yes, unfortunately we need the test executable with debug info to decode the backtrace

**Comment by kbr-scylla** [comment]
> @kbr-scylla please try https://jenkins.scylladb.com/job/scylla-master/job/build/1389/artifact/scylla/build/aarch64/debug/FailingtestExe/recent_entries_map_test

Nope :(
```
?? ??:0
?? ??:0
?? ??:0
?? ??:0
```
and so on.

**Comment by andriipatsula** [comment]
In unit tests:
* `test_remove_least_recent_entries` 
* `test_update_and_remove_least_recent_entries`

I am using the `seastar::manual_clock` to advance the clock. This is the only data structure that somehow related to the `reactor` codebase.

The  `seastar::manual_clock` is used in couple of places and I found interesting comment: https://github.com/scylladb/scylladb/blob/master/test/boost/rate_limiter_test.cc#L100-L104

> // TODO: Workaround for seastar#1072. Calling `manual_clock::advance`
// multiple times and then quitting the test immediately causes
// the test framework to hang. I didn't have the time to debug it, but I
// suspect there are some pending tasks which need to finish before exiting
// from the main test task.

The issue `seastar#1072` is solved long time ago, but the comment was introduced on June 2022.

Just in case: I ran these tests 10'000 times on the `amd64` platform and wan't able to reproduce the problem. I don't have `aarch64` machine to have the same conditions as on the build box.

It 'd be great to run these tests on `aarch64` platform in debug mode and collect a dump.

**Comment by kbr-scylla** [comment]
Ref: https://github.com/scylladb/seastar/issues/1072

**Comment by michoecho** [comment]
> `/jenkins/workspace/scylla-master/build/scylla/seastar/src/core/reactor.cc:2927:25: runtime error: member access within null pointer of type 'seastar::reactor::task_queue'`

I bet it's a duplicate of https://github.com/scylladb/seastar/issues/1514

**Comment by michoecho** [comment]
> Hard to believe it's a bug in seastar reactor unrelated to this test and happened after this particular test, which was added recently...

https://github.com/scylladb/seastar/issues/1514 also happened in the same test. And yet, I think it really is just a coincidence, because the same happened to me many months ago, in a completely different test (https://github.com/scylladb/seastar/issues/1514#issuecomment-1438620388).



> > @kbr-scylla please try https://jenkins.scylladb.com/job/scylla-master/job/build/1389/artifact/scylla/build/aarch64/debug/FailingtestExe/recent_entries_map_test
> 
> Nope :(
> 
> ```
> ?? ??:0
> ?? ??:0
> ?? ??:0
> ?? ??:0
> ```
> 
> and so on.

See b6e427551193fd9a06d12e2c1de832e51a26ac00. In `debug` and `dev` modes, tests are built with shared library seastar now, so you need not only the failing binary, but also the `libseastar.so` and `libseastar_testing.so` for decoding. But jenkins doesn't save them. We should fix that.

**Comment by bhalevy** [comment]
Cc @yaronkaikov @benipeled :point_up:

**Comment by kbr-scylla** [comment]
> In debug and dev modes, tests are built with shared library seastar now, so you need not only the failing binary, but also the libseastar.so and libseastar_testing.so for decoding.

Is this documented anywhere, or just more tribal knowledge to remember? My head is too small for this.

**Comment by kbr-scylla** [comment]
> What do you want documented?

Just the answer to the question: how to decode a backtrace from Scylla or Scylla test.
Previously it was straightforward - get the scylla/test binary, then `seastar-addr2line -e path/to/binary`.
I guess we'll need to update https://github.com/scylladb/scylladb/blob/master/docs/kb/decode-stack-trace.rst

**Comment by michoecho** [comment]
> I guess we'll need to update https://github.com/scylladb/scylladb/blob/master/docs/kb/decode-stack-trace.rst

We won't, because the recent changes only apply to debug/dev tests. Scylla is unaffected.

> Just the answer to the question: how to decode a backtrace from Scylla or Scylla test.
> Previously it was straightforward - get the scylla/test binary, then seastar-addr2line -e path/to/binary.

There's a better way: teach the tests to decode the backtrace on the spot, instead of requiring a human to do that manually. Why haven't we done it already anyway.

**Comment by nyh** [comment]
> > I guess we'll need to update https://github.com/scylladb/scylladb/blob/master/docs/kb/decode-stack-trace.rst
> 
> We won't, because the recent changes only apply to debug/dev tests. Scylla is unaffected.

What about build/dev/scylla - does it not use the shared library?
That's the Scylla I usually debug, not some official release build.

**Comment by michoecho** [comment]
> > > I guess we'll need to update https://github.com/scylladb/scylladb/blob/master/docs/kb/decode-stack-trace.rst
> > 
> > 
> > We won't, because the recent changes only apply to debug/dev tests. Scylla is unaffected.
> 
> What about build/dev/scylla - does it not use the shared library? That's the Scylla I usually debug, not some official release build.

Right, build/dev/scylla and build/debug/scylla do use the shared library. I meant that public docs (https://github.com/scylladb/scylladb/blob/master/docs/kb/decode-stack-trace.rst) are unaffected, because the public uses release builds, and these are unaffected.

Well, build/dev/scylla doesn't contain symbols in the first place, so you can't decode stack traces from it anyway. But if you wanted to decode a stack trace from build/debug/scylla, then yes, the change broke that.

By the way, I forgot to say something important: it isn't enough to have all the binaries, you also need the /proc/PID/maps and tooling to parse it. Simply saving libseastar.so together with the failing test won't help.
+1 point to the idea of automatic trace decoding, I guess.

**Comment by avikivity** [comment]
The rationale was to reduce disk usage. libseastar.so is just a step, we need libscyllacore.so to make significant progress.

**Comment by michoecho** [comment]
@avikivity I could implement automatic (optional, or e.g. crash-only) stack trace decoding, so that the inability to decode debug traces becomes a non-issue.
The way I see it:
1. On startup, seastar starts a side process, which is used solely for decoding stack traces. A side process is needed because you can't do anything interesting besides read() and write() in a signal handler, so all decoding has to happen outside. It has to be started ahead of time, because you can't fork() in a signal handler. The side process communicates with seastar through a pipe.
2. When printing a stack trace and automatic decoding is enabled, let seastar print the stack trace to the side process pipe, instead of stderr. The side process will reply with a decoded stack trace through a second pipe and seastar will print the response to stderr.
3. The side process parses /proc/PID/maps of its parent to find the ELF of the address and its offset therein, and runs that through addr2line.

Refs #12073

**Comment by nyh** [comment]
@michoecho although I like the **option** the shared library, I think it should be a configure.py option and that we don't have to, just yet, make it on by default. We should only make it on by default when the tooling has improved and this option has more pros than cons - and I'm not sure we reached that point yet.

Personally, I almost never build the entire C++ unit test suite (I compile Scylla for Python tests, and perhaps a single C++ unit test, but not all of them) so the entire disk-saving thing is a non-issue for me - but I want to be able to decode backtraces.

Also, how do decode backtraces in test failures in dev/debug builds? Did http://backtrace.scylladb.com/ stop being useful because of this?

**Comment by michoecho** [comment]
> Also, how do decode backtraces in test failures in dev/debug builds?

Decoding dev mode addresses has always been impossible, because they don't produce debug symbols.
Decoding debug mode addresses used to work for all symbols before the libseastar.so patch, and still works for Scylla code and probably most of inlined Seastar code, but stopped working for libseastar.so symbols (non-inlined Seastar code, e.g. reactor stuff, which is all this issue (#13128) is about). Which is a problem.

> Did http://backtrace.scylladb.com/ stop being useful because of this?

Don't we use http://backtrace.scylladb.com/ only for `release` mode binaries?

> @michoecho although I like the option the shared library, I think it should be a configure.py option and that we don't have to, just yet, make it on by default.

I'm not involved in the shared library project, so I'm not the right person to ask about this.

@avikivity @tchaikov We might want to consider disabling libseastar.so by default until we provide some way to decode stack traces which involve it.

**Comment by kbr-scylla** [comment]
> Decoding dev mode addresses has always been impossible, because they don't produce debug symbols.

I recall being able to decode addresses from tests at least, if I compiled the `_g` version of the test.

**Comment by michoecho** [comment]
> > Decoding dev mode addresses has always been impossible, because they don't produce debug symbols.
> 
> I recall being able to decode addresses from tests at least, if I compiled the `_g` version of the test.

Okay, I undersold the dev mode. The decoder can still decode an address without debug info, by falling back to ELF symbols alone. But you get limited information this way — you don't see inlined functions and file/line information, only the names of non-inlined symbols.

**Comment by nyh** [comment]
> > Also, how do decode backtraces in test failures in dev/debug builds?
> 
> Decoding dev mode addresses has always been impossible, because they don't produce debug symbols.

That's not accurate. First of all, like I said, I usually debug the Scylla executable itself (which gets used by many of the test suites), not the C++ unit tests. And the Scylla executable itself does has symbols. Second, while the dev mode C++ unit tests indeed don't have symbols to save space, you can ask to build `sometest_g` it relinks the same object file *without* stripping, and presto, in you have the symbols you need. 

> Decoding debug mode addresses used to work for all symbols before the libseastar.so patch, and still works for Scylla code and probably most of inlined Seastar code, but stopped working for libseastar.so symbols (non-inlined Seastar code, e.g. reactor stuff, which is all this issue (#13128) is about). Which is a problem.
> 
> > Did http://backtrace.scylladb.com/ stop being useful because of this?
> 
> Don't we use http://backtrace.scylladb.com/ only for `release` mode binaries?

Maybe I'm misremembering, but can't we use it to decode backtraces on failed Jenkins tests?
I agree that for debugging customer issues, it will always be release builds.

**Comment by michoecho** [comment]
> And the Scylla executable itself does has symbols.

Right, I've corrected my imprecision in https://github.com/scylladb/scylladb/issues/13128#issuecomment-1471592709. `dev` builds don't have *DWARF* (debug) symbols as I said, but they do have *ELF* symbols as said. With ELF symbols alone you still can decode the names of non-inlined function in the stack trace, but you can't get file/line information. I disregarded pure ELF symbols because I don't get why would someone want to debug a build without debug symbols. `release` compiles only slightly slower than `dev`, but actually has debug symbols — I never saw a reason to use `dev` over `release`.

> Second, while the dev mode C++ unit tests indeed don't have symbols to save space, you can ask to build sometest_g it relinks the same object file without stripping,

Again, sorry for being imprecise about "symbols". `dev` binaries *never* have DWARF symbols, which I focused on. They aren't stripped at link time, because they are not produced at all.
What you are saying is true for ELF symbols. They are produced (they have to, because that's needed for linking), but they get stripped from tests at link time (unless `_g`).

**Comment by bhalevy** [comment]
Hit in: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/85/artifact/testlog/x86_64/debug/boost.recent_entries_map_test.test_update_and_remove_least_recent_entries.1.log

**Comment by bhalevy** [comment]
> > `/jenkins/workspace/scylla-master/build/scylla/seastar/src/core/reactor.cc:2927:25: runtime error: member access within null pointer of type 'seastar::reactor::task_queue'`
> 
> I bet it's a duplicate of [scylladb/seastar#1514](https://github.com/scylladb/seastar/issues/1514)

I managed to reproduce locally with enough repeats (in debug mode).
and got a coredump but gdb seems to get confused with libseastar.so
Cc @tchaikov 

In any case, I agree it's most likely a duplicate of scylladb/seastar#1514

**Comment by michoecho** [comment]
> By the way, I forgot to say something important: it isn't enough to have all the binaries, you also need the /proc/PID/maps and tooling to parse it. Simply saving libseastar.so together with the failing test won't help.

I take that back. It turns out that Seastar already has the machinery to iterate over proc maps, and it's supposed to print DSO addresses not as raw virtual addresses, but as path+offset, which doesn't need additional metadata for decoding. It's just bugged: https://github.com/scylladb/seastar/pull/1558

**Comment by kbr-scylla** [comment]
We don't observe this anymore, https://github.com/scylladb/seastar/issues/1514 was fixed which was most likely the reason.
Closing.
